script.js

```diff
@@ -44,32 +41,108 @@ class ToDoList {
 
-// Joe: update display/DOM
 
-// Joe: update item in toDoList (to set it as "done")
+function updateDisplay(list) {
+
+    const outputElement = document.querySelector('section');
+    const items = list.getAllItems();
+    outputElement.innerHTML = '';
+
+    for (const item of items) {
+
+        const taskContainerEl = document.createElement('div');
 
-// (Stretch goal: hide completed items)
+        const toggleDoneInputEl = document.createElement('input');
+        toggleDoneInputEl.setAttribute('type', 'checkbox');
 
-/*
+        const taskTextContainerEl = document.createElement('span'); // change to div?
 
-    Add, update and delete functions *only* change the array.
-    (They only change the data that we're storing.)
+        // If the item's done flag is set to true:
+        if (item['done'] === true) {
+            // Wrap the text in a strikethrough tag:
+            taskTextContainerEl.innerHTML = `<s>${item['task']}</s>`;
+            // Check the checkbox:
+            toggleDoneInputEl.setAttribute('checked', 'on');
+            // Add the "done" class to this task's text container:
+            taskTextContainerEl.classList.add('done');
+        } else {
+            // If not, then don't wrap the text in a strikethrough tag:
+            taskTextContainerEl.innerHTML = item['task'];
+            /*  (There's no need to remove the "done" class or uncheck the checkbox,
+                as the class isn't assigned by default, and the checkbox is unchecked by default.)
+            */
+
+        }
 
-    The "update display" function *only* modifies the DOM.
-    (It always reads the most up-to-date data that we're storing, and updates the DOM using that.)
+        const deleteButtonEl = document.createElement('input');
+        deleteButtonEl.setAttribute('type', 'button');
+        deleteButtonEl.setAttribute('value', 'X');
 
-    When they've finished changing the array, the add, update, and delete functions
-    will always call the update display function.
+        taskContainerEl.append(toggleDoneInputEl);
+        taskContainerEl.append(taskTextContainerEl);
+        taskContainerEl.append(deleteButtonEl);
 
-    test("addTask() should add a new object to the toDoList array", () => {
-        ...
-    });
+        outputElement.append(taskContainerEl);
+
+        toggleDoneInputEl.addEventListener("input", () => {
+            /*
+                indexOf *seems* to be a safe/reliable way to do this?
+                (i.e. it won't confuse "seemingly-identical" elements.)
+
+                Could we set this const outside of this function (i.e. in the parent scope)?
+                It might cause problems (because if we delete an item's element from the array,
+                its index will change, so that will quickly get out of sync).
+
+                But given we updateDisplay on every event anyway, might it work?
+            */
+            const currentItemIndex = items.indexOf(item);
```

Comment:
`indexOf()` seems a bit complicated. What about using `forEach((element, index) => )` that could give you the index as well?

Reply:
I think finding the index using forEach is possible, but would just be a bit longer? e.g.

``forEach((element, index) => { if (element === item) { return index; })``

``indexOf()`` feels a bit neater here. Unless I've misunderstood?

Reply:
I meant use `forEach()` instead of `for (const item of items)` so you can directly have the index.

Reply:
Ah, I see!

So, although I didn't use ``forEach()``, an earlier version of this code got the item index in the loop, i.e. in the "parent scope" of the eventListener arrow function.

At the time, this caused a problem: if items were deleted, the indexes that the eventListeners referred to were out of date/out of sync with the data stored, and could be wrong.  (The comments on lines 96-100 talk about this.)

So the fix was to get the item's index within the eventListener (which is a callback function, after all -- i.e., when it's called, it'll find the correct/current index, rather than whatever index the item had when the element was added to the document).

I jumped back and forth a couple of times between slightly different approaches, and settled on one which seems to work, and which (I think/hope!) will be the most flexible.

Since settling on this, I didn't actually test to see whether we could get away with getting the item index outside the eventListener callback scope. I think it would have the same problem, but I'm not sure. (I can just test it quickly :)